radix-engine-profiling/src/info_alloc/mod.rs

```diff
@@ -0,0 +1,251 @@
+use std::{
+    alloc::{GlobalAlloc, Layout, System},
+    sync::atomic::{AtomicBool, AtomicIsize, Ordering},
+};
+
+#[global_allocator]
+pub static INFO_ALLOC: InfoAlloc<System> = InfoAlloc::new(System);
+
+thread_local! {
+    static INFO_ALLOC_DATA_TLS: InfoAllocData = InfoAllocData::new();
+}
+
+/// Heap allocations tracker
+///
+/// This allocator information provider can count allocations up to isize::MAX (9_223_372_036_854_775_807),
+/// in case of exceeding this value code will panic with message: 'Value out of range'.
+pub struct InfoAlloc<T: GlobalAlloc> {
+    /// Heap allocator to use, default usage: System
+    allocator: T,
+    /// Determine if allocation data gathering is enabled
+    enabled: AtomicBool,
+}
+
+/// Allocation data stored in Thread Local Store (separate data for each thread)
+pub struct InfoAllocData {
+    /// Sum of bytes allocated during measurements (for reallocations only additional memory is counted, no dealocation is counted)
+    sum_counter: AtomicIsize,
+    /// Current level of allocated bytes (allocation and deallocation are counted, incl. reallocation)
+    current_level: AtomicIsize,
+    /// Maximum level (peak) of allocated bytes (the max of this field and current_level)
+    max_level: AtomicIsize,
```

Comment:
Signed number is needed as the counters can be enabled in any time of program execution, in such scenario:
1. Allocate 10 bytes
2. Start counting allocations (set counters to 0)
3. Free allocated 10 bytes in point 1
this will result in counter current level value equals -10, for unsigned counter we need to have 0.
Then if we will allocate 11 bytes signed counter will have value 1 and unsigned would have 11.

Reply:
Makes sense. Thanks for explaining it.